File: VTI/Services/QuesitoWriterService.swift
import Foundation

struct QuesitoWriterService {
    let repositoryService: RepositoryService

    func save(draft: NewQuesitoDraft) throws -> URL {
        let materia = draft.materia
        let argomento = try sanitizedNameComponent(draft.argomento)
        let titolo = try sanitizedNameComponent(draft.titolo)
        let testoDomanda = draft.testoDomanda.trimmingCharacters(in: .whitespacesAndNewlines)
        let opzioni = draft.opzioni.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }

        guard !argomento.isEmpty, !titolo.isEmpty, !testoDomanda.isEmpty else {
            throw AppError.ioError("Completa materia, argomento, titolo e testo domanda.")
        }
        guard opzioni.count == 5, opzioni.allSatisfy({ !$0.isEmpty }) else {
            throw AppError.ioError("Il quesito deve avere esattamente 5 opzioni, tutte compilate.")
        }

        let fileURL = try buildOutputFileURL(materia: materia, argomento: argomento, titolo: titolo)
        let content = makeContent(question: testoDomanda, options: opzioni, correctAnswer: draft.rispostaCorretta)
        try content.write(to: fileURL, atomically: true, encoding: .utf8)
        return fileURL
    }

    func importTXT(
        sourceURL: URL,
        materia: Materia,
        argomento: String,
        titolo: String
    ) throws -> URL {
        let cleanedArgomento = try sanitizedNameComponent(argomento)
        let cleanedTitolo = try sanitizedNameComponent(titolo)
        guard !cleanedArgomento.isEmpty, !cleanedTitolo.isEmpty else {
            throw AppError.ioError("Inserisci materia, argomento e titolo per l'import.")
        }

        let raw = try String(contentsOf: sourceURL, encoding: .utf8)
        guard !raw.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw AppError.ioError("Il file TXT selezionato e vuoto.")
        }

        let fileURL = try buildOutputFileURL(materia: materia, argomento: cleanedArgomento, titolo: cleanedTitolo)
        try raw.write(to: fileURL, atomically: true, encoding: .utf8)
        return fileURL
    }

    func makeContent(question: String, options: [String], correctAnswer: Int) -> String {
        let optionLines = options
            .map { "    \\item \($0)" }
            .joined(separator: "\n")

        return """
        \\itemdomanda{\(question)}{
        \(optionLines)
        }

        Risposta Corretta: \(correctAnswer)
        """
    }

    /// Blocco `\itemdomanda`… come nel corpo dei test `.tex`, senza riga soluzione (stesso criterio del parser).
    func latexBlockForPreview(draft: NewQuesitoDraft) -> String? {
        let argomento = draft.argomento.trimmingCharacters(in: .whitespacesAndNewlines)
        let titolo = draft.titolo.trimmingCharacters(in: .whitespacesAndNewlines)
        let testoDomanda = draft.testoDomanda.trimmingCharacters(in: .whitespacesAndNewlines)
        let opzioni = draft.opzioni.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }

        guard !argomento.isEmpty, !titolo.isEmpty, !testoDomanda.isEmpty,
              opzioni.count == 5, opzioni.allSatisfy({ !$0.isEmpty }) else {
            return nil
        }

        let full = makeContent(question: testoDomanda, options: opzioni, correctAnswer: draft.rispostaCorretta)
        return stripRispostaLine(from: full)
    }

    private func stripRispostaLine(from text: String) -> String {
        text
            .split(separator: "\n", omittingEmptySubsequences: false)
            .filter { !$0.trimmingCharacters(in: .whitespaces).hasPrefix("Risposta Corretta:") }
            .joined(separator: "\n")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func nextAvailableIndex(materia: Materia, argomento: String, titolo: String, folder: URL) throws -> Int {
        let prefix = "\(materia.rawValue) - \(argomento) - \(titolo) #"
        let files = try FileManager.default.contentsOfDirectory(at: folder, includingPropertiesForKeys: nil)
        let matches = files
            .map(\.lastPathComponent)
            .filter { $0.hasPrefix(prefix) && $0.hasSuffix(".txt") }
            .compactMap { fileName -> Int? in
                let trimmed = fileName.replacingOccurrences(of: "\(prefix)", with: "").replacingOccurrences(of: ".txt", with: "")
                return Int(trimmed)
            }

        return (matches.max() ?? -1) + 1
    }

    private func buildOutputFileURL(materia: Materia, argomento: String, titolo: String) throws -> URL {
        let folder = try repositoryService.quesitiFolder(for: materia)
        let nextIndex = try nextAvailableIndex(materia: materia, argomento: argomento, titolo: titolo, folder: folder)
        let fileName = "\(materia.rawValue) - \(argomento) - \(titolo) #\(nextIndex).txt"
        return folder.appendingPathComponent(fileName)
    }

    private func sanitizedNameComponent(_ value: String) throws -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        let invalid = CharacterSet(charactersIn: "/:\\?%*|\"<>")
        guard trimmed.rangeOfCharacter(from: invalid) == nil else {
            throw AppError.invalidFileNameComponent(value)
        }
        return trimmed.replacingOccurrences(of: "\n", with: " ")
    }
}

File: VTI.Windows/src/VTI.Core/QuesitoWriterService.cs
namespace VTI.Core;

public sealed class QuesitoWriterService
{
    private readonly RepositoryService _repository;

    public QuesitoWriterService(RepositoryService repository) => _repository = repository;

    public string Save(NewQuesitoDraft draft)
    {
        var materia = draft.Materia;
        var argomento = SanitizedNameComponent(draft.Argomento);
        var titolo = SanitizedNameComponent(draft.Titolo);
        var testoDomanda = draft.TestoDomanda.Trim();
        var opzioni = draft.Opzioni.Select(o => o.Trim()).ToList();
        if (string.IsNullOrEmpty(argomento) || string.IsNullOrEmpty(titolo) || string.IsNullOrEmpty(testoDomanda))
            throw AppException.Io("Completa materia, argomento, titolo e testo domanda.");
        if (opzioni.Count != 5 || opzioni.Any(string.IsNullOrEmpty))
            throw AppException.Io("Il quesito deve avere esattamente 5 opzioni, tutte compilate.");
        var filePath = BuildOutputFilePath(materia, argomento, titolo);
        var content = MakeContent(testoDomanda, opzioni, draft.RispostaCorretta);
        File.WriteAllText(filePath, content);
        return filePath;
    }

    public string ImportTxt(string sourcePath, Materia materia, string argomento, string titolo)
    {
        var cleanedArgomento = SanitizedNameComponent(argomento);
        var cleanedTitolo = SanitizedNameComponent(titolo);
        if (string.IsNullOrEmpty(cleanedArgomento) || string.IsNullOrEmpty(cleanedTitolo))
            throw AppException.Io("Inserisci materia, argomento e titolo per l'import.");
        var raw = File.ReadAllText(sourcePath);
        if (string.IsNullOrWhiteSpace(raw))
            throw AppException.Io("Il file TXT selezionato e vuoto.");
        var filePath = BuildOutputFilePath(materia, cleanedArgomento, cleanedTitolo);
        File.WriteAllText(filePath, raw);
        return filePath;
    }

    public static string MakeContent(string question, IReadOnlyList<string> options, int correctAnswer)
    {
        var optionLines = string.Join("\n", options.Select(o => $"    \\item {o}"));
        return $"""
            \itemdomanda{{{question}}}{{
            {optionLines}
            }}

            Risposta Corretta: {correctAnswer}
            """;
    }

    public string? LatexBlockForPreview(NewQuesitoDraft draft)
    {
        var argomento = draft.Argomento.Trim();
        var titolo = draft.Titolo.Trim();
        var testoDomanda = draft.TestoDomanda.Trim();
        var opzioni = draft.Opzioni.Select(o => o.Trim()).ToList();
        if (string.IsNullOrEmpty(argomento) || string.IsNullOrEmpty(titolo) || string.IsNullOrEmpty(testoDomanda)
            || opzioni.Count != 5 || opzioni.Any(string.IsNullOrEmpty))
            return null;
        var full = MakeContent(testoDomanda, opzioni, draft.RispostaCorretta);
        return StripRispostaLine(full);
    }

    private static string StripRispostaLine(string text) =>
        string.Join("\n", text.Split('\n')
            .Where(line => !line.Trim().StartsWith("Risposta Corretta:", StringComparison.OrdinalIgnoreCase)))
            .Trim();

    private string BuildOutputFilePath(Materia materia, string argomento, string titolo)
    {
        var folder = _repository.QuesitiFolder(materia);
        var nextIndex = NextAvailableIndex(materia, argomento, titolo, folder);
        var fileName = $"{materia.RawValue} - {argomento} - {titolo} #{nextIndex}.txt";
        return Path.Combine(folder, fileName);
    }

    private static int NextAvailableIndex(Materia materia, string argomento, string titolo, string folder)
    {
        var prefix = $"{materia.RawValue} - {argomento} - {titolo} #";
        if (!Directory.Exists(folder)) return 0;
        var max = -1;
        foreach (var f in Directory.EnumerateFiles(folder, "*.txt"))
        {
            var name = Path.GetFileName(f);
            if (!name.StartsWith(prefix, StringComparison.Ordinal) || !name.EndsWith(".txt", StringComparison.Ordinal))
                continue;
            var mid = name.Substring(prefix.Length);
            mid = mid[..mid.IndexOf('.')];
            if (int.TryParse(mid, out var n))
                max = Math.Max(max, n);
        }
        return max + 1;
    }

    private static string SanitizedNameComponent(string value)
    {
        var trimmed = value.Trim();
        const string invalid = "/:\\?%*|\"<>";
        if (trimmed.IndexOfAny(invalid.ToCharArray()) >= 0)
            throw AppException.InvalidFileNameComponent(value);
        return trimmed.Replace("\n", " ", StringComparison.Ordinal);
    }
}
